validation/src/pipeline.rs

```diff
@@ -0,0 +1,375 @@
+		if encoded_pov.len() as u64 > max_size {
+			return Err(Error::BlockDataTooBig { size: encoded_pov.len() as _, max_size });
+		}
+	}
+
+	let hash = BlakeTwo256::hash(encoded_pov);
+	if hash != collation.pov_block_hash {
+		return Err(Error::PoVHashMismatch(collation.pov_block_hash, hash));
+	}
+
+	if let Err(()) = collation.check_signature() {
+		return Err(Error::InvalidCollatorSignature);
+	}
+
+	Ok(())
+}
+
+struct ExternalitiesInner {
+	upward: Vec<UpwardMessage>,
+	fees_charged: Balance,
+	free_balance: Balance,
+	fee_schedule: FeeSchedule,
+}
+
+impl wasm_executor::Externalities for ExternalitiesInner {
+	fn post_upward_message(&mut self, message: UpwardMessage) -> Result<(), String> {
+		self.apply_message_fee(message.data.len())?;
+
+		self.upward.push(message);
+
+		Ok(())
+	}
+}
+
+impl ExternalitiesInner {
+	fn new(free_balance: Balance, fee_schedule: FeeSchedule) -> Self {
+		Self {
+			free_balance,
+			fee_schedule,
+			fees_charged: 0,
+			upward: Vec::new(),
+		}
+	}
+
+	fn apply_message_fee(&mut self, message_len: usize) -> Result<(), String> {
+		let fee = self.fee_schedule.compute_fee(message_len);
+		let new_fees_charged = self.fees_charged.saturating_add(fee);
+		if new_fees_charged > self.free_balance {
+			Err("could not cover fee.".into())
+		} else {
+			self.fees_charged = new_fees_charged;
+			Ok(())
+		}
+	}
+
+	// Returns the noted outputs of execution so far - upward messages and balances.
+	fn outputs(self) -> (Vec<UpwardMessage>, Balance) {
+		(self.upward, self.fees_charged)
+	}
+}
+
+#[derive(Clone)]
+struct Externalities(Arc<Mutex<ExternalitiesInner>>);
+
+impl Externalities {
+	fn new(free_balance: Balance, fee_schedule: FeeSchedule) -> Self {
+		Self(Arc::new(Mutex::new(
+			ExternalitiesInner::new(free_balance, fee_schedule)
+		)))
+	}
+}
+
+impl wasm_executor::Externalities for Externalities {
+	fn post_upward_message(&mut self, message: UpwardMessage) -> Result<(), String> {
+		self.0.lock().post_upward_message(message)
+	}
+}
+
+/// Data from a fully-outputted validation of a parachain candidate. This contains
+/// all outputs and commitments of the validation as well as all additional data to make available.
+pub struct FullOutput {
+	/// Data about the candidate to keep available in the network.
+	pub available_data: AvailableData,
+	/// Commitments issued alongside the candidate to be placed on-chain.
+	pub commitments: CandidateCommitments,
+	/// All erasure-chunks associated with the available data. Each validator
+	/// should keep their chunk (by index). Other chunks do not need to be
+	/// kept available long-term, but should be distributed to other validators.
+	pub erasure_chunks: Vec<ErasureChunk>,
+}
+
+impl FullOutput {
+	/// Check consistency of the outputs produced by the validation pipeline against
+	/// data contained within a candidate receipt.
+	pub fn check_consistency(&self, receipt: &AbridgedCandidateReceipt) -> Result<(), Error> {
+		if self.commitments == receipt.commitments {
```

Comment:
Shouldn't this be the other way around?

Reply:
why does it matter? it's an equality check.